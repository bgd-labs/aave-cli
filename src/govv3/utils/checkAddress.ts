import { findObjectPaths } from 'find-object-paths';
import { Address, getAddress } from 'viem';
import * as addresses from '@bgd-labs/aave-address-book';

/**
 * Checks if address is listed on address-book
 * @param value
 * @param chainId
 * @returns string[] found paths to address-book addresses
 */
export function isKnownAddress(value: Address, chainId: number): string[] | void {
  // glob imports have no object properties
  // therefore we recreate the object via spread & remove addresses unrelated to the chain we are checking
  const transformedAddresses = Object.keys(addresses).reduce((acc, key) => {
    if ((addresses as any)[key].CHAIN_ID == chainId) acc[key] = { ...(addresses as any)[key] };
    return acc;
  }, {} as { [key: string]: any });
  // while on address book we have checksummed addresses tenderly returns non check summed
  // therefore we checksum the needle to have exact matches
  const results = findObjectPaths(transformedAddresses, { value: getAddress(value) });
  if (typeof results === 'string') return [results];
  return results;
}

export function findPayloadsController(chainId: number): Address | void {
  const key = Object.keys(addresses).find(
    (key) => addresses[key].CHAIN_ID === chainId && addresses[key].PAYLOADS_CONTROLLER
  );
  if (key) return addresses[key].PAYLOADS_CONTROLLER;
}
