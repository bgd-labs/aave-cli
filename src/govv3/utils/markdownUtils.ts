import { Hex, PublicClient } from 'viem';
import { CheckResult } from '../checks/types';

export function boolToMarkdown(value: boolean) {
  if (value) return `:white_check_mark:`;
  return `:sos:`;
}

/**
 * Turns a plaintext address into a link to etherscan page of that address
 * @param address to be linked
 * @param code whether to link to the code tab
 */
export function toAddressLink(address: Hex, md: boolean, client?: PublicClient): string {
  if (!client) return address;
  const link = `${client.chain?.blockExplorers?.default.url}/address/${address}`;
  if (md) return toMarkdownLink(link, address);
  return link;
}

/**
 * Turns a plaintext address into a link to etherscan page of that address
 * @param address to be linked
 * @param code whether to link to the code tab
 */
export function toTxLink(txn: Hex, md: boolean, client?: PublicClient): string {
  if (!client) return txn;
  const link = `${client.chain?.blockExplorers?.default.url}/tx/${txn}`;
  if (md) return toMarkdownLink(link, txn);
  return link;
}

export function toMarkdownLink(link: string, title?: any) {
  return `[${title || link}](${link})`;
}

export function renderCheckResult(check: { name: string }, result: CheckResult) {
  let response = `### Check: ${check.name} ${boolToMarkdown(!result.errors.length)}\n\n`;
  if (result.errors.length) response += `#### Errors\n\n${result.errors.join('\n')}\n\n`;
  if (result.warnings.length) response += `#### Warnings\n\n${result.warnings.join('\n')}\n\n`;
  if (result.info.length) response += `#### Info\n\n${result.info.join('\n')}\n\n`;
  return response;
}

export function renderUnixTime(time: number) {
  return new Date(time * 1000).toLocaleString('en-GB', { timeZone: 'UTC' });
}
