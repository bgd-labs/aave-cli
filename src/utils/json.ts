import JSONbig from "json-bigint";
import fs from "fs";
import path from "path";

export function readJson(filePath: string) {
  const content = fs.readFileSync(path.join(process.cwd(), filePath));
  return JSON.parse(
    JSON.stringify(JSONbig({ storeAsString: true }).parse(content))
  );
}
