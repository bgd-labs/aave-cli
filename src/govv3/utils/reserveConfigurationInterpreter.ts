import * as pools from '@bgd-labs/aave-address-book';
import {getBits} from '../../utils/storageSlots';

export function getDecodedReserveData(contractAddress: string, data?: bigint) {
  if (!data) return data;
  if (
    [
      pools.AaveV2EthereumAMM.POOL,
      pools.AaveV2Ethereum.POOL,
      pools.AaveV2Polygon.POOL,
      pools.AaveV2Avalanche.POOL,
    ]
      .map((address) => address.toLowerCase())
      .includes(contractAddress.toLowerCase())
  )
    return decodeReserveDataV2(data);
  return decodeReserveDataV3(data);
}

function decodeReserveDataV2(data: bigint) {
  const ltv = getBits(data, 0n, 15n);
  const liquidationThreshold = getBits(data, 16n, 31n);
  const liquidationBonus = getBits(data, 32n, 47n);
  const decimals = getBits(data, 48n, 55n);
  const active = Number(getBits(data, 56n, 56n));
  const frozen = Number(getBits(data, 57n, 57n));
  const borrowingEnabled = Number(getBits(data, 58n, 58n));
  const stableBorrowingEnabled = Number(getBits(data, 59n, 59n));
  const reserveFactor = getBits(data, 64n, 79n);
  return {
    ltv,
    liquidationThreshold,
    liquidationBonus,
    decimals,
    active: !!active,
    frozen: !!frozen,
    borrowingEnabled: !!borrowingEnabled,
    stableBorrowingEnabled: !!stableBorrowingEnabled,
    reserveFactor,
  };
}

function decodeReserveDataV3(data: bigint) {
  const ltv = getBits(data, 0n, 15n);
  const liquidationThreshold = getBits(data, 16n, 31n);
  const liquidationBonus = getBits(data, 32n, 47n);
  const decimals = getBits(data, 48n, 55n);
  const active = Number(getBits(data, 56n, 56n));
  const frozen = Number(getBits(data, 57n, 57n));
  const borrowingEnabled = Number(getBits(data, 58n, 58n));
  const stableRateBorrowingEnabled = Number(getBits(data, 59n, 59n));
  const paused = Number(getBits(data, 60n, 60n));
  const borrowingInIsolation = Number(getBits(data, 61n, 61n));
  const siloedBorrowingEnabled = Number(getBits(data, 62n, 62n));
  const flashloaningEnabled = Number(getBits(data, 63n, 63n));
  const reserveFactor = getBits(data, 64n, 79n);
  const borrowCap = getBits(data, 80n, 115n);
  const supplyCap = getBits(data, 116n, 151n);
  const liquidationProtocolFee = getBits(data, 152n, 167n);
  const eModeCategory = getBits(data, 168n, 175n);
  const unbackedMintCap = getBits(data, 176n, 211n);
  const debtCeiling = getBits(data, 212n, 251n);

  return {
    ltv,
    liquidationThreshold,
    liquidationBonus,
    decimals,
    active: !!active,
    frozen: !!frozen,
    borrowingEnabled: !!borrowingEnabled,
    stableRateBorrowingEnabled,
    paused,
    borrowingInIsolation,
    reserveFactor,
    borrowCap,
    supplyCap,
    liquidationProtocolFee,
    eModeCategory,
    unbackedMintCap,
    debtCeiling,
    siloedBorrowingEnabled: !!siloedBorrowingEnabled,
    flashloaningEnabled: !!flashloaningEnabled,
  };
}
