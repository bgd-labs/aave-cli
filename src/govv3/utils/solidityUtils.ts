import { getAddress } from 'viem';
import { TenderlySimulationResponse } from '../../utils/tenderlyClient';

// --- Helper methods ---
/**
 * @notice Given a Tenderly contract object, generates a descriptive human-friendly name for that contract
 * @param contract Tenderly contract object to generate name from
 */
export function getContractName(contracts: TenderlySimulationResponse['contracts'] = [], address: string): string {
  // lower-case comparison as strict equality would fail on child contracts which are lower-cased when returned from tenderly
  const contract = contracts.find((c) => c.address.toLowerCase() === address.toLowerCase());
  if (!contract) return `unknown contract name at \`${getAddress(address)}\``;
  let contractName = contract?.contract_name;

  // If the contract is a token, include the full token name. This is useful in cases where the
  // token is a proxy, so the contract name doesn't give much useful information
  if (contract?.token_data?.name) contractName += ` (${contract?.token_data?.name})`;

  // if the contract is a proxy include it's child
  if (contract.standards?.includes('eip1967') && contract.child_contracts?.[0].address) {
    return `${contractName} at \`${getAddress(contract.address)}\` with implementation ${getContractName(
      contracts,
      contract.child_contracts?.[0].address
    )}`;
  }

  // Lastly, append the contract address and save it off
  return `${contractName} at \`${getAddress(contract.address)}\``;
}
