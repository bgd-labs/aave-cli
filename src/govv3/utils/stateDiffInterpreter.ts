import { Hex, PublicClient, getContract } from 'viem';
import { tenderlyDeepDiff } from './tenderlyDeepDiff';
import { ERC20_ABI } from '../abis/ERC20';
import * as pools from '@bgd-labs/aave-address-book';
import { getBits } from '../../utils/storageSlots';

export async function interpretStateChange(
  contractAddress: string,
  name: string = '',
  original: Record<string, any>,
  dirty: Record<string, any>,
  key: Hex,
  publicClient: PublicClient
) {
  if (name === '_reserves' && (original.configuration.data || dirty.configuration.data))
    return await reserveConfigurationChanged(contractAddress, original, dirty, key, publicClient);
  return undefined;
}

async function reserveConfigurationChanged(
  contractAddress: string,
  original: Record<string, any>,
  dirty: Record<string, any>,
  key: Hex,
  publicClient: PublicClient
) {
  const configurationBefore = getDecodedReserveData(contractAddress, original.configuration.data);
  const configurationAfter = getDecodedReserveData(contractAddress, dirty.configuration.data);
  let symbol = 'unknown';
  try {
    const erc20Contract = getContract({ publicClient, address: key, abi: ERC20_ABI });
    symbol = await erc20Contract.read.symbol();
  } catch (e) {}
  // const symbol =
  return `# decoded configuration.data for key \`${key}\` (symbol: ${symbol})
  ${tenderlyDeepDiff(configurationBefore, configurationAfter, 'configuration.data')}`;
}

function getDecodedReserveData(contractAddress: string, data?: bigint) {
  if (!data) return {};
  if (
    [pools.AaveV2EthereumAMM.POOL, pools.AaveV2Ethereum.POOL, pools.AaveV2Polygon.POOL, pools.AaveV2Avalanche.POOL]
      .map((address) => address.toLowerCase())
      .includes(contractAddress.toLowerCase())
  )
    return decodeReserveDataV2(data);
  return decodeReserveDataV3(data);
}

function decodeReserveDataV2(data: bigint) {
  const ltv = getBits(data, 0n, 15n);
  const liquidationThreshold = getBits(data, 16n, 31n);
  const liquidationBonus = getBits(data, 32n, 47n);
  const decimals = getBits(data, 48n, 55n);
  const active = Number(getBits(data, 56n, 56n));
  const frozen = Number(getBits(data, 57n, 57n));
  const borrowingEnabled = Number(getBits(data, 58n, 58n));
  const reserveFactor = getBits(data, 64n, 79n);
  return {
    ltv,
    liquidationThreshold,
    liquidationBonus,
    decimals,
    active: !!active,
    frozen: !!frozen,
    borrowingEnabled: !!borrowingEnabled,
    reserveFactor,
  };
}

function decodeReserveDataV3(data: bigint) {
  const ltv = getBits(data, 0n, 15n);
  const liquidationThreshold = getBits(data, 16n, 31n);
  const liquidationBonus = getBits(data, 32n, 47n);
  const decimals = getBits(data, 48n, 55n);
  const active = Number(getBits(data, 56n, 56n));
  const frozen = Number(getBits(data, 57n, 57n));
  const borrowingEnabled = Number(getBits(data, 58n, 58n));
  const stableRateBorrowingEnabled = Number(getBits(data, 59n, 59n));
  const paused = Number(getBits(data, 60n, 60n));
  const borrowingInIsolation = Number(getBits(data, 61n, 61n));
  const siloedBorrowingEnabled = Number(getBits(data, 62n, 62n));
  const flashloaningEnabled = Number(getBits(data, 63n, 63n));
  const reserveFactor = getBits(data, 64n, 79n);
  const borrowCap = getBits(data, 80n, 115n);
  const supplyCap = getBits(data, 116n, 151n);
  const liquidationProtocolFee = getBits(data, 152n, 167n);
  const eModeCategory = getBits(data, 168n, 175n);
  const unbackedMintCap = getBits(data, 176n, 211n);
  const debtCeiling = getBits(data, 212n, 251n);

  return {
    ltv,
    liquidationThreshold,
    liquidationBonus,
    decimals,
    active: !!active,
    frozen: !!frozen,
    borrowingEnabled: !!borrowingEnabled,
    stableRateBorrowingEnabled,
    paused,
    borrowingInIsolation,
    reserveFactor,
    borrowCap,
    supplyCap,
    liquidationProtocolFee,
    eModeCategory,
    unbackedMintCap,
    debtCeiling,
    siloedBorrowingEnabled: !!siloedBorrowingEnabled,
    flashloaningEnabled: !!flashloaningEnabled,
  };
}
