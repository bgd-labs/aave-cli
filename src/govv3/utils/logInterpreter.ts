import {type Client, type Hex, formatUnits} from 'viem';
import type {Input} from '../../utils/tenderlyClient';
import {findAsset} from './checkAddress';
import {formatNumberString, prettifyNumber} from './markdownUtils';

// events emitted typically on the erc20
const tokenAmountEvents = ['Transfer', 'Approval', 'Burn', 'Mint', 'BalanceTransfer'];

// events emitted on the pool
const reserveEvents = ['Withdraw', 'Supply', 'Deposit'];

export async function interpretLog(
  client: Client,
  address: Hex,
  name: string | null,
  inputs: Input[],
) {
  if (name && tokenAmountEvents.includes(name)) {
    // fields formatted by the asset decimal
    const decimalFieldNames = ['value', 'amount', 'wad'];
    for (const name of decimalFieldNames) {
      const valueIndex = inputs.findIndex((i) => i.soltype!.name === name);
      if (valueIndex !== -1) {
        const asset = await findAsset(client, address);
        if (asset) {
          inputs[valueIndex].value = prettifyNumber({
            value: inputs[valueIndex].value as string,
            decimals: asset.decimals,
          });
        }
      }
    }
  }
  if (name && reserveEvents.includes(name)) {
    const valueIndex = inputs.findIndex((i) => i.soltype!.name === 'amount');
    const reserveIndex = inputs.findIndex((i) => i.soltype!.name === 'reserve');
    if (valueIndex !== -1 && reserveIndex !== -1) {
      const asset = await findAsset(client, inputs[reserveIndex].value as Hex);
      if (asset) {
        inputs[valueIndex].value = prettifyNumber({
          value: inputs[valueIndex].value as string,
          decimals: asset.decimals,
        });
      }
    }
  }
  if (name && ['Mint', 'ReserveDataUpdated', 'Burn'].includes(name)) {
    // fields formatted by the asset decimal
    const decimalFieldNames = ['liquidityIndex', 'variableBorrowIndex', 'index'];
    for (const name of decimalFieldNames) {
      const valueIndex = inputs.findIndex((i) => i.soltype!.name === name);
      if (valueIndex !== -1) {
        inputs[valueIndex].value = prettifyNumber({
          value: inputs[valueIndex].value as string,
          decimals: 27,
        });
      }
    }
  }
  const parsedInputs = inputs
    ?.map(
      (i) =>
        `${i.soltype!.name}: ${typeof i.value === 'object' ? JSON.stringify(i.value) : i.value}`,
    )
    .join(', ');
  return `  - \`${name}(${parsedInputs || ''})\``;
}
