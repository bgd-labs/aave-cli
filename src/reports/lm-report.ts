import {Client, type Hex} from 'viem';
import {getClient} from '../utils/getClient';
import {formatTimestamp, toAddressLink, addAssetSymbolWithLink} from '../govv3/utils/markdownUtils';
import {LiquidityMiningSnapshot} from './snapshot-types';

export async function generateLiquidityMiningReport(snapshot: LiquidityMiningSnapshot) {
  let content = '';
  content += `# Liquidity Mining Report\n\n`;
  content += '<br/>\n\n';

  for (const key in snapshot) {
    content += '\n\n';
    const object = snapshot[key];

    content += `Asset: ${await addAssetSymbolWithLink(getClient(object.chainId) as Client, object.asset as Hex)}\n`;
    content += `Reward: ${await addAssetSymbolWithLink(getClient(object.chainId) as Client, object.reward as Hex)}\n`;
    content += `Type: ${object.type}\n`;
    if (object.index) content += `Index: ${object.index}\n`;

    if (object.newTransferStrategy) content += `_Transfer Strategy changed to: ${toAddressLink(object.newTransferStrategy as Hex, true, getClient(object.chainId))}_\n`;
    if (object.newRewardOracle) content += `_Reward Oracle changed to: ${toAddressLink(object.newRewardOracle as Hex, true, getClient(object.chainId))}_\n`;

    if (object.newDistributionEnd || object.newEmission) {
      content += `|  | prev value | newValue |\n`;
      content += `| - | - | - |\n`;
      if (object.newEmission && object.oldEmission) content += `| emissionPerSecond | ${object.oldEmission} | ${object.newEmission} |\n`;
      if (object.newDistributionEnd && object.oldDistributionEnd) content += `| emissionPerSecond | ${object.oldDistributionEnd} (${formatTimestamp(object.oldDistributionEnd)}) | ${object.newDistributionEnd} (${formatTimestamp(object.newDistributionEnd)}) |\n`;
    }

    content += '<br/>\n\n';
    content += '---\n';
  }

  return content;
}
