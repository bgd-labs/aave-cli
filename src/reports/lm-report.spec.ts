import {describe, expect, it} from 'vitest';
import {readJsonFile} from '../utils/json';
import {generateLiquidityMiningReport} from './lm-report';

describe('liquidity mining report', () => {
  it(
    'should generate a well formatted liquidity mining report',
    async () => {
      const snapshot = readJsonFile('/src/reports/mocks/liquidityMining.json');
      const content = await generateLiquidityMiningReport(snapshot);
      expect(content).toMatchSnapshot();
    },
    {timeout: 30000},
  );
});
