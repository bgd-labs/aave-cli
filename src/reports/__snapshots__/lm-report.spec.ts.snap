// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`liquidity mining report > should generate a well formatted liquidity mining report 1`] = `
"# Liquidity Mining Report

<br/>



Asset: [variableDebtPolWMATIC](https://polygonscan.com/address/0x4a1c3aD6Ed28a636ee1751C69071f6be75DEb8B8)
Reward: [MaticX](https://polygonscan.com/address/0xfa68FB4628DFF1028CFEc22b4162FCcd0d45efb6)
Type: LM_UPDATE
Index: 2109876369033151
_Transfer Strategy changed to: [0x53F57eAAD604307889D87b747Fc67ea9DE430B01](https://polygonscan.com/address/0x53F57eAAD604307889D87b747Fc67ea9DE430B01)_
_Reward Oracle changed to: [0xB0A72A5e454890e9715e059e8df8582a6B383DE3](https://polygonscan.com/address/0xB0A72A5e454890e9715e059e8df8582a6B383DE3)_
|  | prev value | newValue |
| - | - | - |
| emissionPerSecond | 1695570156 (24 Sept 2023) | 1739992061 (19 Feb 2025) |
<br/>

---
"
`;
